blockscout-ens/bens-logic/src/protocols/protocoler.rs

```diff
@@ -288,17 +290,114 @@ impl Protocoler {
         name: &str,
         network_id: i64,
         maybe_filter: Option<NonEmpty<String>>,
+    ) -> Result<Vec<DomainNameOnProtocol>, ProtocolError> {
+        let tlds = self
+            .networks
+            .get(&network_id)
+            .ok_or_else(|| ProtocolError::NetworkNotFound(network_id))?
+            .use_protocols
+            .iter()
+            .filter_map(|protocol_name| {
+                self.protocols
+                    .get(protocol_name)
+                    .map(|protocol| protocol.info.tld_list.iter().cloned())
+            })
+            .flatten()
+            .collect::<Vec<Tld>>();
+
+        if name.contains('.') {
+            let direct = self.names_options_in_network_exact(name, network_id, maybe_filter)?;
+            if direct.is_empty() {
+                Err(ProtocolError::InvalidName {
+                    name: name.to_string(),
+                    reason: "No matching protocols for given TLD".to_string(),
+                })
+            } else {
+                Ok(direct.into_iter().take(MAX_NAMES_LIMIT).collect())
+            }
+        } else {
+            let all_names_with_protocols: Vec<_> = tlds
+                .into_iter()
+                .map(|tld| format!("{}.{}", name, tld.0))
+                .flat_map(|name_with_tld| {
+                    self.names_options_in_network_with_suggestions(
+                        &name_with_tld,
+                        network_id,
+                        maybe_filter.clone(),
+                    )
+                    .unwrap_or_default()
+                })
+                .take(MAX_NAMES_LIMIT)
+                .collect();
+
+            if all_names_with_protocols.is_empty() {
+                Err(ProtocolError::InvalidName {
+                    name: name.to_string(),
+                    reason: "No valid TLDs".to_string(),
+                })
+            } else {
+                Ok(all_names_with_protocols)
+            }
+        }
+    }
+
+    pub fn names_options_in_network_exact(
```

Comment:
maybe
```suggestion
   fn names_options_in_network_exact(
```
if it's not expected to be used outside